File: Editorial/774_Minimize_Max_Distance_to_Gas_Station/Java/4_Approach_4_Approach_3_5_Preprocessing_Heap_Greedy_Accepted.java
import java.util.*;
import static java.lang.Math.*;

class Solution {
    public double minmaxGasDist(int[] stations, int K) {
        int N = stations.length;
        double distUpper = (stations[N - 1] - stations[0]) / (double)(K + 1);
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(b[0], a[0]));

        for (int i = 0; i < N - 1; i++) {
            double interval = stations[i + 1] - stations[i];
            int ki = (int) Math.floor(interval / distUpper);
            double actualDist = interval / (ki + 1);
            K -= ki;
            pq.add(new double[]{actualDist, interval, ki});
        }

        while (K-- > 0) {
            double[] top = pq.poll();
            double dist = top[0];
            double interval = top[1];
            int ki = (int) top[2] + 1;
            pq.add(new double[]{interval / (ki + 1), interval, ki});
        }

        return pq.peek()[0];
    }
}

File: Editorial/774_Minimize_Max_Distance_to_Gas_Station/Python3/4_Approach_4_Approach_3_5_Preprocessing_Heap_Greedy_Accepted.py
from heapq import heappush, heappop
from math import floor

class Solution:
    def minmaxGasDist(self, stations, k):
        n = len(stations)

        # Step 1: Compute an upper bound for distance
        dist_upper = (stations[-1] - stations[0]) / (k + 1)

        # Step 2: Conservative allocation
        heap = []  # (-distance, interval_length, current_station_count)
        for i in range(n - 1):
            interval = stations[i + 1] - stations[i]
            ki = floor(interval / dist_upper)
            actual_dist = interval / (ki + 1)
            k -= ki
            heappush(heap, (-actual_dist, interval, ki))

        # Step 3: Allocate remaining stations greedily
        for _ in range(k):
            neg_dist, interval, ki = heappop(heap)
            ki += 1
            new_dist = interval / (ki + 1)
            heappush(heap, (-new_dist, interval, ki))

        # Step 4: The largest remaining interval
        return -heap[0][0]
